### PR TITLE
Move activation logic for SearchOptions.WHOLE_WORDS to FindReplaceLogic

### PR DESCRIPTION
**Pull request by Wittmaxi**
when "whole words" is selected but a find-string was entered that is not a whole word, the dialog/overlay will disable the option for "whole words". However, the dialog/overlay will keep the selection of the button in it's internal state. By passing the findString to this method, we can make sure that we don't try to perform a whole-words search for a string that isn't a whole word.

See https://github.com/eclipse-platform/eclipse.platform.ui/pull/1192#discussion_r1548564603

Extracted from
https://github.com/eclipse-platform/eclipse.platform.ui/pull/1192

### CODE REVIEW

**Comment by Wittmaxi** [comment]
Random failing tests are documented [here](https://github.com/eclipse-platform/eclipse.platform.ui/issues/1583) and [here](https://github.com/eclipse-platform/eclipse.platform.ui/issues/1517)

**Review by HeikoKlare — CHANGES_REQUESTED** [review]
This introduces a duplication of the check for the "whole word" option being available in the `FindReplaceDialog` and the `FindReplaceLogic`. In addition, the implementations are different, as the one added to `FindReplaceLogic` only checks for the find string not containing a whitespace, while `FindReplaceDialog` has sophisticated logic for checking the find string for being a valid "whole word" in the `isWord()` method.

Proposal: Move the functionality for checking availability of the "whole word" option completely to the `FindReplaceLogic` and provided it to the dialog. E.g., the  isWholeWordSearchAvailableAndActive(String)` method could be made part of the `IFindReplaceLogic` API. Then, in `FindReplaceDialog` the business logic for making the decision on the availability of the option can be removed and replaced by a call to that method of the `FindReplaceLogic`.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java [hunk 1]
```diff
@@ -166,4 +166,15 @@ public interface IFindReplaceLogic {
 	 */
 	public IFindReplaceTarget getTarget();
 
+	/**
+	 * Returns <code>true</code> if searching should be restricted to entire words,
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]:
```suggestion
	 * Returns <code>true</code> if searching can be restricted to entire words,
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java [hunk 2]
```diff
@@ -166,4 +166,15 @@ public interface IFindReplaceLogic {
...
+	 * <code>false</code> if not. This is the case if the respective checkbox is
+	 * turned on, regex is off, and the checkbox is enabled, i.e. the current find
+	 * string is an entire word.
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]: There are no check boxes at the level of the business logic.
```suggestion
	 * <code>false</code> if not. Searching for whole words requires the given find string to be an entire word and the regex search option to be disabled.
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java [hunk 3]
```diff
@@ -166,4 +166,15 @@ public interface IFindReplaceLogic {
...
+	 *
+	 * @param findString the string that is currently being searched for.
+	 * @return <code>true</code> if the search is restricted to whole words
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]:
```suggestion
	 * @return <code>true</code> if the search can be restricted to whole words
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java [hunk 4]
```diff
@@ -114,24 +114,45 @@ private void resetStatus() {
...
+
+		return !isRegExSearchAvailableAndActive();
+	}
+
+	boolean isWholeWordSearchAvailableAndActive(String findString) {
+		if (isWholeWordSearchAvailable(findString)) {
+			return isActive(SearchOptions.WHOLE_WORD);
+		}
+		return false;
+	}
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: This method at least needs to be private and the body can be simplified to:
```java
return isWholeWordSearchAvailable(findString) && isActive(SearchOptions.WHOLE_WORD);
```
It may also be sufficient to just inline this condition at its consumers.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java [hunk 5]
```diff
@@ -114,24 +114,45 @@ private void resetStatus() {
...
+	 * Tests whether each character in the given string is a letter.
 	 *
-	 * @return <code>true</code> if the search is restricted to whole words
+	 * @param str the string to check
+	 * @return <code>true</code> if the given string is a word
+	 * @since 3.0
 	 */
-	private boolean isWholeWordSearchAvailableAndActive() {
-		return isActive(SearchOptions.WHOLE_WORD) && !isRegExSearchAvailableAndActive();
+	private boolean isWord(String str) {
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]:
```suggestion
	private static boolean isWord(String str) {
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java [hunk 6]
```diff
@@ -114,24 +114,45 @@ private void resetStatus() {
...
+		if (str == null || str.isEmpty())
+			return false;
+
+		for (int i = 0; i < str.length(); i++) {
+			if (!Character.isJavaIdentifierPart(str.charAt(i)))
+				return false;
+		}
+		return true;
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: This can be simplified:
```suggestion
return str != null && str.chars().allMatch(Character::isJavaIdentifierPart);
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java [hunk 7]
```diff
@@ -497,11 +516,10 @@ private int findIndex(String findString, int startPosition) {
 	public int findAndSelect(int offset, String findString) {
 		if (target instanceof IFindReplaceTargetExtension3)
 			return ((IFindReplaceTargetExtension3) target).findAndSelect(offset, findString,
-					isActive(SearchOptions.FORWARD),
-					isActive(SearchOptions.CASE_SENSITIVE), isWholeWordSearchAvailableAndActive(), isActive(SearchOptions.REGEX));
+					isActive(SearchOptions.FORWARD), isActive(SearchOptions.CASE_SENSITIVE),
+					isWholeWordSearchAvailable(findString), isActive(SearchOptions.REGEX));
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]:
```suggestion
					isWholeWordSearchAvailableAndActive(findString), isActive(SearchOptions.REGEX));
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java [hunk 8]
```diff
@@ -497,11 +516,10 @@ private int findIndex(String findString, int startPosition) {
...
 		return target.findAndSelect(offset, findString, isActive(SearchOptions.FORWARD),
-				isActive(SearchOptions.CASE_SENSITIVE),
-				isWholeWordSearchAvailableAndActive());
+				isActive(SearchOptions.CASE_SENSITIVE), isWholeWordSearchAvailable(findString));
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]:
```suggestion
				isActive(SearchOptions.CASE_SENSITIVE), isWholeWordSearchAvailableAndActive(findString));
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java [hunk 9]
```diff
@@ -1065,7 +1065,7 @@ private void updateButtonState(boolean disableReplace) {
 			boolean isSelectionGoodForReplace = selectionString != "" //$NON-NLS-1$
 					|| !isRegExSearchAvailableAndActive;
 
-			fWholeWordCheckBox.setEnabled(isWord(str) && !isRegExSearchAvailableAndActive);
+			fWholeWordCheckBox.setEnabled(findReplaceLogic.isWholeWordSearchAvailable(getFindString()));
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java:1068]: I would expect all places setting the enablement of the whole-word checkbox to be adapted accordingly. In particular, I would expect line 1181 to be adapted as well:
```java
fWholeWordCheckBox.setEnabled(!findReplaceLogic.isRegExSearchAvailableAndActive());
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java [hunk 10]
```diff
@@ -114,24 +114,45 @@ private void resetStatus() {
 		status = null;
 	}
 
+	@Override
+	public boolean isWholeWordSearchAvailable(String findString) {
+		if (!isWord(findString)) { // $NON-NLS-1$
+			return false;
+		}
+
+		return !isRegExSearchAvailableAndActive();
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: Shorter and easier to read:
```suggestion
		return isWord(findString) && !isRegExSearchAvailableAndActive();
```

**Review by HeikoKlare — COMMENTED** [review]
Please also add a regression test for the changed behavior / new API.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java (hunk 10 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java (hunk 9 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java:1068]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java (hunk 8 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java (hunk 7 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java (hunk 6 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java (hunk 5 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java (hunk 4 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java]: I have decided that it is reasonable to bundle this logic into a function so as to not introduce added complexity into the functions down below.

Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java (hunk 3 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java (hunk 2 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java (hunk 1 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/a0b224321663e63aa99f60da4a2527f993d68f7d

**Comment by Wittmaxi** [comment]
Random failing test documented [here](https://github.com/eclipse-platform/eclipse.platform.ui/issues/867)

**Comment by Wittmaxi** [comment]
I have extracted a requested fix for a similar coupling-problem for the IncrementalSearch into this PR: https://github.com/eclipse-platform/eclipse.platform.ui/pull/1805
These PRs are not separate from eachother and each provide testing for their own functionality

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java [hunk 11]
```diff
@@ -1280,10 +1263,8 @@ private void setupFindReplaceLogic() {
 		activateInFindReplaceLogicIf(SearchOptions.FORWARD, fForwardRadioButton.getSelection());
 		activateInFindReplaceLogicIf(SearchOptions.CASE_SENSITIVE, fCaseCheckBox.getSelection());
 		activateInFindReplaceLogicIf(SearchOptions.REGEX, fIsRegExCheckBox.getSelection());
-		activateInFindReplaceLogicIf(SearchOptions.WHOLE_WORD,
-				fWholeWordCheckBox.getEnabled() && fWholeWordCheckBox.getSelection());
-		activateInFindReplaceLogicIf(SearchOptions.INCREMENTAL,
-				fIncrementalCheckBox.getEnabled() && fIncrementalCheckBox.getSelection());
+		activateInFindReplaceLogicIf(SearchOptions.WHOLE_WORD, fWholeWordCheckBox.getSelection());
+		activateInFindReplaceLogicIf(SearchOptions.INCREMENTAL, fIncrementalCheckBox.getSelection());
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java]: This change is part of the extraction to #1805 and must not be part of this PR.

**Review by HeikoKlare — COMMENTED** [review]
Works as expected. The activation of the "whole word" options is now also reapplied as expected when reopening the find/replace dialog of it the checkbox was not enabled when closing it (because the find string was not a single word when closing the dialog).

There only need for a minor change as there is still one line change that needs to be extracted to #1805.

**Comment by HeikoKlare** [comment]
And can you please adapt the commit headline to be more expressed, in particular by refering to the find/replace functionality? Something like "Move activation logic for whole word search option to FindReplaceLogic"

**Comment by Wittmaxi** [comment]
@HeikoKlare I have addressed your comments in https://github.com/eclipse-platform/eclipse.platform.ui/commit/33cc2984f3f11f33d32b7dbb6f65110922db40df

**Review by HeikoKlare — APPROVED** [review]
Perfect, thank you for updating, and thank you for the improvement in general, @Wittmaxi!

**Comment by akurtakov** [comment]
Is it possible that this change causes the failing test in jdt.ui master? https://ci.eclipse.org/jdt/job/eclipse.jdt.ui-github/job/master/76/testReport/junit/AutomatedSuite%20org.eclipse.jdt.ui.tests.search.SearchTest%20org.eclipse.jdt.ui.tests.search/NLSSearchTest/testBug247012_1/
```
org.eclipse.swt.SWTException: Widget is disposed
	at org.eclipse.swt.SWT.error(SWT.java:4922)
	at org.eclipse.swt.SWT.error(SWT.java:4837)
	at org.eclipse.swt.SWT.error(SWT.java:4808)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:597)
	at org.eclipse.swt.widgets.Widget.checkWidget(Widget.java:512)
	at org.eclipse.swt.widgets.Widget.getData(Widget.java:623)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:842)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:672)
	at org.eclipse.ui.internal.e4.compatibility.ActionBars.updateActionBars(ActionBars.java:81)
	at org.eclipse.ui.part.PageBookView.showPageRec(PageBookView.java:929)
	at org.eclipse.ui.part.PageBookView.partActivated(PageBookView.java:698)
	at org.eclipse.search2.internal.ui.SearchView.partActivated(SearchView.java:333)
	at org.eclipse.search2.internal.ui.SearchView.internalShowSearchPage(SearchView.java:422)
	at org.eclipse.search2.internal.ui.SearchView.showSearchResult(SearchView.java:393)
	at org.eclipse.search2.internal.ui.SearchViewManager.showNewSearchQuery(SearchViewManager.java:74)
	at org.eclipse.search2.internal.ui.SearchViewManager$1.queryAdded(SearchViewManager.java:48)
	at org.eclipse.search2.internal.ui.QueryManager.fireAdded(QueryManager.java:96)
	at org.eclipse.search2.internal.ui.QueryManager.addQuery(QueryManager.java:73)
	at org.eclipse.search2.internal.ui.InternalSearchUI.addQuery(InternalSearchUI.java:311)
	at org.eclipse.search2.internal.ui.InternalSearchUI.runSearchInForeground(InternalSearchUI.java:207)
	at org.eclipse.search.ui.NewSearchUI.runQueryInForeground(NewSearchUI.java:190)
	at org.eclipse.search.ui.NewSearchUI.runQueryInForeground(NewSearchUI.java:161)
	at org.eclipse.jdt.ui.tests.search.NLSSearchTestHelper.searchProblems(NLSSearchTestHelper.java:59)
```

**Comment by iloveeclipse** [comment]
No, it is https://github.com/eclipse-platform/eclipse.platform.ui/issues/1668
